### ISSUE TITLE
API is always returning "You are not allowed to view this content" and then the requested content

### ISSUE THREAD

**Issue by pbksol** [thorsten/phpMyFAQ#1620]
**Describe the bug**
After upgrading FAQ from V3.0.0b3 to RC1 the Api returns always an answer like this:

`{"error":"You are not allowed to view this content."}[{"date":"2019-12-24T13:14:00+0100","question":"`

As you can see, the error is returned, but the API also returns the requested content right afterwards if you check the raw response.

As an example, this is the response after a getVersion:

`{"error":"You are not allowed to view this content."}{"version":"3.0.0-RC"}` 

There's something broken. The response is always "error" regardless a valid login but contains the response in every case.

**Expected behavior**
No error message

**Comment by thorsten** [comment]
Do you use the old v1 endpoints or the new v2 ones?

**Comment by pbksol** [comment]
`/api.php?action=getVersion` – that's the only endpoints shown in https://www.phpmyfaq.de/docs/3.0#8.2

**Comment by thorsten** [comment]
Did you secured your FAQ?

**Comment by pbksol** [comment]
Yes.

<img width="441" alt="Bildschirmfoto 2020-01-12 um 18 35 01" src="https://user-images.githubusercontent.com/8238713/72222930-49d93a80-356a-11ea-8bca-5c68c6c980ba.png">

I did revert everything. But there's no session cookie returned after login – did check the session in my history and there's nothing after upgrading to 3RC. I was a bit led on the wrong track because the API did return the result.

And the Login did respond with a 200 and that was it. That narrows it down to an unsuccessful login. The problem with the API returning all contents even if there's no login must be older. I tried an API request on 3b and this version sends the same response as cited in my first post. Error + Result what it shouldn't.

Did you change something for the login in 3RC?

**Comment by thorsten** [comment]
Not for the login, but I changed a lot of stuff in the API for the RC. I see some issues now and I'll fix them.

**Comment by thorsten** [comment]
@pbksol can you check these two fixes?

**Comment by thorsten** [comment]
@pbksol It's working on my development environment using Postman, I created a Postman collection for you:

```json
{
	"info": {
		"_postman_id": "d99608cb-0958-41e5-b9d3-e2004f4f6395",
		"name": "Login via API and fetch version",
		"schema": "https://schema.getpostman.com/json/collection/v2.1.0/collection.json"
	},
	"item": [
		{
			"name": "Login",
			"request": {
				"method": "POST",
				"header": [],
				"url": {
					"raw": "http://localhost:8080/api/v2.0/login?user=username&password=password",
					"protocol": "http",
					"host": [
						"localhost"
					],
					"port": "8080",
					"path": [
						"api",
						"v2.0",
						"login"
					],
					"query": [
						{
							"key": "user",
							"value": "username"
						},
						{
							"key": "password",
							"value": "password"
						}
					]
				}
			},
			"response": []
		},
		{
			"name": "Version",
			"request": {
				"method": "GET",
				"header": [],
				"url": {
					"raw": ""
				}
			},
			"response": []
		}
	],
	"protocolProfileBehavior": {}
}
```

**Comment by pbksol** [comment]
Nope. Login is still broken. That's the full response to POST https://xxxxx.invalid/api.php?action=login with faqusername=xxxxxx&faqpassword=xxxxxxxxx in the body.

No login, no error. Just no JSON object is returned.

`HTTP/1.1 200 OK
Date: Mon, 13 Jan 2020 16:47:34 GMT
Server: Apache
Expires: Thu, 07 Apr 1977 14:47:00 GMT
Last-Modified: Mon, 13 Jan 2020 16:47:34 GMT
Cache-Control: no-store, no-cache, must-revalidate
Cache-Control: post-check=0, pre-check=0
Pragma: no-cache
Vary: Negotiate,Accept,Accept-Encoding
Connection: close
Transfer-Encoding: chunked
Content-Type: application/json
`

The fixes do repair the invalid JSON answers containing the content. That part is fixed.

But no login. I reverted back to b3 and login works again.

Is there any API V2 documentation available?

**Comment by thorsten** [comment]
The API v2 documentation is here: https://github.com/thorsten/phpMyFAQ/blob/3.0/API.md

**Comment by pbksol** [comment]
Thanks.

Login via API v2 does work. Login via API v1 does not and returns no JSON and no Cookie.

**Comment by thorsten** [comment]
The request works for v1, too. Please see the example in the Postman collection:

```json
{
	"info": {
		"_postman_id": "8eafce31-c9ee-47a1-b340-9c06d1f68660",
		"name": "v1: Login via API and fetch version",
		"schema": "https://schema.getpostman.com/json/collection/v2.1.0/collection.json"
	},
	"item": [
		{
			"name": "Login",
			"request": {
				"method": "POST",
				"header": [],
				"body": {
					"mode": "formdata",
					"formdata": [
						{
							"key": "username",
							"value": "username",
							"type": "text"
						},
						{
							"key": "password",
							"value": "password",
							"type": "text"
						}
					]
				},
				"url": {
					"raw": "http://localhost:8080/api/login",
					"protocol": "http",
					"host": [
						"localhost"
					],
					"port": "8080",
					"path": [
						"api",
						"login"
					]
				}
			},
			"response": []
		},
		{
			"name": "Version",
			"request": {
				"method": "GET",
				"header": [],
				"url": {
					"raw": ""
				}
			},
			"response": []
		}
	],
	"protocolProfileBehavior": {}
}
```

**Comment by thorsten** [comment]
So, put username/password as form-data in the POST body.

**Comment by pbksol** [comment]
So the documentation under https://www.phpmyfaq.de/docs/3.0#8.2 is no longer valid.

Would be really, really great to somewhat communicate such changes (which nobody expects when changing from Beta to a RC). Would have made things a lot easier ;-)

**Comment by thorsten** [comment]
Well, you're right, I'll remove the v1 API docs from the 3.0 docs as I want the users to use the new v2 one, sorry for the late changes between the latest beta and the RC.

**Comment by thorsten** [comment]
The v2 API is much more RESTful and provides correct HTTP status codes.

**Comment by pbksol** [comment]
Yep. V2 is much more matured. It's a it of work to switch everything from V1 to V2 so it's good to have both running in parallel. I'll migrate everything over to V2 as soon as I've some time.

Thanks for all the effort you put into this project.

**Comment by pbksol** [comment]
Just for Info… I reinstalled everything again, replaced the files from #1620, changed the login parameters and login works. But there's something else badly broken. The app accessing the FAQ is not able to parse any data. I give up at this point. The FAQ is just backend for something else and is only accessed via api. Reverted everything back to B3 and it's back to normal. I'll stick with B3.

**Comment by thorsten** [comment]
What do you mean by „not parsing any data“?

**Comment by pbksol** [comment]
It seems as if it's struggling with the responses. But I have no time to dig any deeper at this point. The whole API connector is optimized to process responses from prior 3RC versions of phpmyfaq and a bit complex. It might be something minor, but I can't say for sure. B3 is working which is fine so far. Due to time limitations I'll have to delay any further investigations for at least a month.

**Comment by thorsten** [comment]
Okay, please create a new issue if you find the reason what's happening with your installation.

I updated the 3.0 docs here: https://www.phpmyfaq.de/docs/3.0#8.2
